# SQL

Компания решила купить крупный сервис для чтения книг по подписке. 

**Цель исследования:** необходимо проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.



**Задачи:**

- Посчитать, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитать количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определить автора с самой высокой средней оценкой книг — учитывая только книги с 50 и более оценками;
- Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

**Доступ к базе данных**

[Подключение к базе данных .pdf](https://s3.us-west-2.amazonaws.com/secure.notion-static.com/bac02c0e-9489-4458-acff-4cd0ad912b46/instructions_for_connecting_to_the_database.pdf?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Content-Sha256=UNSIGNED-PAYLOAD&X-Amz-Credential=AKIAT73L2G45EIPT3X45%2F20220611%2Fus-west-2%2Fs3%2Faws4_request&X-Amz-Date=20220611T143234Z&X-Amz-Expires=86400&X-Amz-Signature=5d28268d1e4a3e5788a8ebba7beadd8a9dfb1c07e60342e2f7fe28c5261fddfa&X-Amz-SignedHeaders=host&response-content-disposition=filename%20%3D%22instructions_for_connecting_to_the_database.pdf%22&x-id=GetObject)

## Подключение к базе данных,  импорт библиотек

In [11]:
import pandas as pd
from sqlalchemy import create_engine

In [12]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

## Исследование таблиц

### Таблица books

**Содержит данные о книгах:**

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

In [13]:
query = '''
SELECT 
    *
FROM books
'''

In [14]:
pd.io.sql.read_sql(query, con = engine) 

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

В базе содержится информация о 1000 книг.

### Таблица authors:

**Содержит данные об авторах:**

- `author_id` — идентификатор автора;
- `author` — имя автора.

In [15]:
query = '''
SELECT 
    *
FROM authors
'''

In [16]:
pd.io.sql.read_sql(query, con = engine)

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

Всего в таблице представлена информация о 636 авторах работ.

### Таблица publishers:

**Содержит данные об издательствах:**

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

In [17]:
query = '''
SELECT
    *
FROM publishers    
'''

In [18]:
pd.io.sql.read_sql(query, con = engine)

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

Количество уникальных издательств - 340.

### Таблица ratings:

**Содержит данные о пользовательских оценках книг:**

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

In [19]:
query = '''
SELECT
    *
FROM ratings
'''

In [20]:
pd.io.sql.read_sql(query, con = engine)

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

1000 книг пользователи оценили 6456 оценками. Вероятнее всего оценка может быть от 1 до 5.

### Таблица reviews:

**Содержит данные о пользовательских обзорах на книги:**

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

In [21]:
query = '''
SELECT
    *
FROM reviews
'''

In [22]:
pd.io.sql.read_sql(query, con = engine)

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

На 1000 различных книг в таблице есть 2793 развернутых отзыва.

## Задания

### *Посчитаем, сколько книг вышло после 1 января 2000 года*

In [23]:
pd.io.sql.read_sql('''
SELECT 
    COUNT(DISTINCT book_id) AS cnt
FROM 
    books
WHERE
    publication_date >= '2000-01-01'
''', con = engine)

cnt
0  821

**Ответ:** с начала 2000 года вышла бОльшая часть книг, которая имеется на платформе, а именно 819 из 1000.


### *Для каждой книги посчитаем количество обзоров и среднюю оценку*

In [31]:
pd.io.sql.read_sql('''
SELECT 
    books.title AS title,
    COUNT(DISTINCT reviews.text) AS cnt_reviews,
    AVG(ratings.rating) AS avg_rating
FROM 
    books
    LEFT JOIN reviews ON books.book_id = reviews.book_id
    LEFT JOIN ratings ON books.book_id = ratings.book_id  

GROUP BY
    books.title
ORDER BY
    cnt_reviews DESC, 
    avg_rating DESC;   
''', con = engine)

title  cnt_reviews  \
0                                  Memoirs of a Geisha            8   
1                              Twilight (Twilight  #1)            7   
2    Harry Potter and the Prisoner of Azkaban (Harr...            6   
3    Harry Potter and the Chamber of Secrets (Harry...            6   
4                                       The Book Thief            6   
..                                                 ...          ...   
994  Disney's Beauty and the Beast (A Little Golden...            0   
995                          Essential Tales and Poems            0   
996                               Leonardo's Notebooks            0   
997    Anne Rice's The Vampire Lestat: A Graphic Novel            0   
998                            The Natural Way to Draw            0   

     avg_rating  
0      4.138462  
1      3.662500  
2      4.414634  
3      4.287500  
4      4.264151  
..          ...  
994    4.000000  
995    4.000000  
996    4.000000  
997    3.666667  
998    3.000000  

[999 rows x 3 columns]

- Больше всего отзывов написано на книгу "Мемуары Гейши" - 8, со средним рейтингом пользователей - 4.1;
- На втором месте "Сумерки" с 7-ью обзорами и средней оценкой 3.7;
- На третьем месте по количеству обзоров и среднего рейтинга - книга "Гарри Поттер и узник Азкабана".

### Определим издательство, которое выпустило наибольшее число книг толще 50 страниц (исключая из анализа брошюры)

In [25]:
pd.io.sql.read_sql('''
SELECT 
   publishers.publisher AS publisher,
   COUNT(book_id) AS cnt_books
FROM 
    books
    JOIN publishers ON books.publisher_id = publishers.publisher_id
WHERE num_pages >50 
GROUP BY publisher
ORDER BY cnt_books DESC
LIMIT 1;   
''', con = engine)

publisher  cnt_books
0  Penguin Books         42

Больше всего книг выпустило издательство "Penguin Books".

### Определим автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [26]:
pd.io.sql.read_sql('''
SELECT
    authors.author,
    AVG(subquery_ratings.avg_rating) AS avg_rating
FROM
    books
    JOIN(
        SELECT 
            book_id,
            AVG(rating) AS avg_rating,
            COUNT(rating_id) AS rating_cnt
        FROM
            ratings
        GROUP BY
            book_id
    ) AS subquery_ratings ON subquery_ratings.book_id = books.book_id
    JOIN authors ON authors.author_id = books.author_id
WHERE
    rating_cnt >= 50
GROUP BY
    author
ORDER BY
    avg_rating DESC
LIMIT 1
'''
, con = engine)

author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.283844

Автором с лучшей средней оценкой на книги является Джоан Роулинг - автор книг вселенной Гарри Поттера. Через слэш указана Mary GrandPré - художнк иллюстраций.

### *Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок.*

In [27]:
pd.io.sql.read_sql('''
SELECT 
    AVG(count) AS avg_review    
FROM
    (SELECT
         COUNT(text)
    FROM 
        reviews
        JOIN(
            SELECT 
                username, 
                COUNT(rating_id) as rating_cnt 
            FROM ratings 
            GROUP BY username) AS rating
         ON rating.username = reviews.username 
WHERE rating_cnt > 50
GROUP BY reviews.username) AS count;                                  
'''
, con = engine)    

avg_review
0   24.333333

Среднее количество обзоров от активных пользователй - 24.

## Общий вывод:

В ходе данного исследования была проанализирована база данных крупного сервиса для чтения книг по подписке. В ней содержится информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. В резулате анализа получены следующие результаты:

* В базе данных нового сервиса содержится информация о тысяче книг, что является небольшим количеством для крупных сервисев.
* В БД содержится информация о 636 авторах, 340 издательствах, 6456 оценках книг и о 2793 обзорах на книги.
* На сервисе доступно 819 книг, которые выпущены начиная с 2000-го года.
* Максимальное количество обзоров на книгу - 8.
* Издательство, которое выпустило наибольшее число книг - это Penguin Books. Оно выпустило 42 книги из 1000.
* Самая высокая средняя оценка у автора Джоан Роулинг. Средняя оценка составляет 4.28 балла.
* В среднем активные пользователи написали более 24 подробных отзывов каждый.

Относительно рекоменций по поводу покупки сервиса можно сказать, что в сервисе довольно много активных пользователей, но книг относительно мало. Существует множество сервисев с тысячами книг, которые будут более интересны книголюбам, из-за большого разнообразия. Но помимо варианта расширить библиотеку, есть и другие. Можно сделать цену на подписку ниже рыночной или заинтересовать пользователей хорошей рекламой с книгами набирающими популярность. Бесплатный пробный период тоже довольно частая практика среди приложений. 

В общем и целом можно сказать, что все зависит от того сколько стоит такой сервис, сколько в него хотят вкладывать и какую прибыль и когда хотят начать получать.